### PR TITLE
issue #217, JSON parser with try/catch

### PR DESCRIPTION
**Pull request by chinuy**
Turns out the testing is a little bit tricky. Please verify before merge. Thanks.

### CODE REVIEW

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

File: test/mysql.test.js [hunk 1]
```diff
@@ -81,6 +81,23 @@ describe('mysql', function() {
     });
   });
 
+  it('should return null if JSON.parse fail', function(done) {
+    Post.create({title: 'a', history: {a: 1, b: 'b'}}, function(err, post) {
+      should.not.exist(err);
```
Comment (superkhau) [inline test/mysql.test.js]: `if (err) return done(err);` for better stack traces

File: test/mysql.test.js [hunk 2]
```diff
@@ -81,6 +81,23 @@ describe('mysql', function() {
     });
   });
 
+  it('should return null if JSON.parse fail', function(done) {
```
Comment (superkhau) [inline test/mysql.test.js]: it('returns null if JSON parsing fails'...

File: test/mysql.test.js [hunk 3]
```diff
@@ -81,6 +81,23 @@ describe('mysql', function() {
...
+    Post.create({title: 'a', history: {a: 1, b: 'b'}}, function(err, post) {
+      should.not.exist(err);
+      
+      // break the content
+      var sql = 'UPDATE ' + global.getConfig().database + '.PostWithDefaultId SET history = \'{corrupted,\' where title=\'a\'';
```
Comment (superkhau) [inline test/mysql.test.js]: var statement = ...

File: test/mysql.test.js [hunk 4]
```diff
@@ -81,6 +81,23 @@ describe('mysql', function() {
...
+      db.connector.query(sql, function(err, res) {
```
Comment (superkhau) [inline test/mysql.test.js]: sql -> statement

File: test/mysql.test.js [hunk 5]
```diff
@@ -81,6 +81,23 @@ describe('mysql', function() {
...
+        Post.findById(post.id, function(err, p) {
+          p.id.should.be.equal(post.id);
+          should(p.history).be.null;
+
```
Comment (superkhau) [inline test/mysql.test.js]: remove empty space

File: test/mysql.test.js [hunk 6]
```diff
@@ -81,6 +81,23 @@ describe('mysql', function() {
...
   });
 
+  it('should return null if JSON.parse fail', function(done) {
+    Post.create({title: 'a', history: {a: 1, b: 'b'}}, function(err, post) {
+      should.not.exist(err);
+      
+      // break the content
+      var sql = 'UPDATE ' + global.getConfig().database + '.PostWithDefaultId SET history = \'{corrupted,\' where title=\'a\'';
+      db.connector.query(sql, function(err, res) {
+        Post.findById(post.id, function(err, p) {
```
Comment (superkhau) [inline test/mysql.test.js]: if (err) return done(err);
p -> post

File: test/mysql.test.js [hunk 7]
```diff
@@ -81,6 +81,23 @@ describe('mysql', function() {
...
+          p.id.should.be.equal(post.id);
+          should(p.history).be.null;
```
Comment (superkhau) [inline test/mysql.test.js]: should.not.exist(post.history);

**Comment by chinuy** [comment]
test case enhanced with @superkhau's comments

File: test/mysql.test.js [hunk 8]
```diff
@@ -81,6 +81,23 @@ describe('mysql', function() {
     });
   });
 
+  it('return null if JSON parsing fails', function(done) {
```
Comment (superkhau) [inline test/mysql.test.js]: plural -> returns

File: test/mysql.test.js [hunk 9]
```diff
@@ -81,6 +81,23 @@ describe('mysql', function() {
...
+    Post.create({title: 'a', history: {a: 1, b: 'b'}}, function(err, post) {
+      if (err) return done(err);
+      should.not.exist(err);
```
Comment (superkhau) [inline test/mysql.test.js]: remove should.not.exist(err); line as if (err) return done(err) does it all.

**Comment by chinuy** [comment]
Do you ask for returning error in the `lib/mysql.js`?
Assigning `null` is only because the `MySQL.prototype.fromColumnValue()` function needs to return a `val`.  Current try/catch mechanism only prevents it from crashing main program.

**Comment by superkhau** [comment]
@chinuy Depends how you see the original code -- If we want to keep things sync, we should be try/catching from the caller(s) and not in the code via coercing to null. In this case throwing is already correct. If we were to treat it async, we would return 'err' normally via callback. Those are the two choices I see, what are your thoughts?

**Comment by chinuy** [comment]
@superkhau  Agree. This can be many use cases, depends on users. I believe try/catch is necessary. Just where to put it, and what kind of message should deliver to the user.

A more complicated case might be:
a Model.find() function is call, with just one row containing corrupted data. What's the best thing to return? bare in mind that any error will block all following logic, which isn't always good for a production system.

To me, as an user, I hope the system:

1. return a `null` to me so my system can keep running
2. log such event in my log file so I can inspect the problem in the future

**Comment by chinuy** [comment]
I want to withdraw a little. Maybe throwing out an error to user is a good way, as long as the user can distinguish whether err is critical. Your thought?

**Comment by superkhau** [comment]
My thought is to refactor this code to support callback style so we can return any error to using expected nodejs patterns. function(err, ...) so the user can handle in the callback or log to some file in your use case.

@bajtos Thoughts?

**Comment by bajtos** [comment]
I think there are two parts of this problem:

(1) How to signal the problem to the user? The first solution that comes to my mind is that `Post.findById` in the newly added tests should fail with an error. This is follows the Node.js pattern - if there is something wrong, then fail with an error.

This has a caveat though: the code calling `findById` cannot easily fix the problem, i.e. the app user will keep seeing 500 Internal Error responses when querying the malformed record. The current proposal, where we convert invalid field to `null`, may be better in this light. But then it hides the problem, which is not great either.

(2) Assuming we decide that `findById` should return an error when a JSON value cannot be parsed: I think converting `fromColumnValue` to callback-style is overkill and it will only complicate the code base. IMO, we should move the try-catch block to a different place (possibly to multiple places - everywhere we call `fromColumnValue`).

AFAICT, `fromColumnValue` is used by `fromRow` ([source](https://github.com/strongloop/loopback-connector/blob/c6a68ec5d8c94f293f04eab6986e69113797ebc3/lib/sql.js#L1168)) which is in turn used by `all` ([source](https://github.com/strongloop/loopback-connector/blob/c6a68ec5d8c94f293f04eab6986e69113797ebc3/lib/sql.js#L1200)).

I think this issue affects other database connectors beyond MySQL, possibly all SQL connectors, therefore it's a good idea to fix the issue in the shared connector base (see the links I posted).

**Comment by superkhau** [comment]
>IMO, we should move the try-catch block to a different place (possibly to multiple places - >everywhere we call fromColumnValue).
>
>AFAICT, fromColumnValue is used by fromRow (source) which is in turn used by all (source).
>
>I think this issue affects other database connectors beyond MySQL, possibly all SQL connectors, >therefore it's a good idea to fix the issue in the shared connector base (see the links I posted).

I also think this is the correct solution. Keep the code sync and let the callers handle the try/catches and return errors accordingly. Which is pain to refac because it touches all the places you mentioned. ;( @chinuy Thats the solution. ;) Would you like to try?

**Comment by bajtos** [comment]
> Which is pain to refac because it touches all the places you mentioned. ;( 

Not really. I believe we only need to fix [SQLConnector.prototype.all](https://github.com/strongloop/loopback-connector/blob/c6a68ec5d8c94f293f04eab6986e69113797ebc3/lib/sql.js#L1200):

```js
try {
  var objs = data.map(function(obj) {
    return self.fromRow(model, obj)
  });
} catch (err) {
  return cb(err);
}
```

**Comment by chinuy** [comment]
how about using `async`, so rows with correct data can be preserved, while rows with malformed data will return null

such as:
```js
async.map(objs, function(obj, done) {
  try {
    done(null, self.fromRow(model, obj));
  } catch(err) {
    done(err, null);
  }
}, function(err, res) {

});
```

**Comment by superkhau** [comment]
@chinuy I agree with @bajtos's comment at https://github.com/strongloop/loopback-connector-mysql/pull/218#issuecomment-265697182 to modify only `prototype.all` and return the error to the user (ie. for you to handle/change to null there).

**Comment by chinuy** [comment]
https://github.com/chinuy/loopback-connector/commit/678500c2f24f81d5ab8fef9cac44566d60a25b31

I try to fix this problem based on @bajtos 's comment. However, the callback involved `filter.include` seems tangled.

**Comment by superkhau** [comment]
@chinuy Let's close this PR as rejected and you can submit a new PR with the suggested fixed. Feel free to ping @bajtos or me directly when you need another review.

**Pull request closed without merging**